### PR TITLE
Validate extension dependencies in the packaging phase of the runtime artifact

### PR DESCRIPTION
**Pull request by aloubyansky**
This change adds a validation logic into the Maven plugin that generates the `quarkus-extension.properties` descriptor for the runtime extension artifacts.

Specifically, it is making sure that:
1) the deployment artifact depends on the runtime artifact (although that could possibly be relaxed by enhancing the resolver a bit);
2) if a runtime artifact depends on another extension's runtime artifact then the deployment artifact depends on the corresponding deployment artifact of the other extension.

The validation is enabled by default and, as a consequence, the PR includes fixes extensions that failed the validation.
The validation can be disabled by adding `-DskipExtensionValidation` to the command line.

### CODE REVIEW

**Comment by aloubyansky** [comment]
BTW, I had to move tests from `extensions/smallrye-context-propagation/deployment` to `integration-tests/smallrye-context-propagation` to avoid cyclic build dependencies.

**Comment by aloubyansky** [comment]
As an example, here is an actual error from the CI
```
[INFO] --- quarkus-bootstrap-maven-plugin:999-SNAPSHOT:extension-descriptor (generate-extension-descriptor) @ quarkus-reactive-datasource ---
[ERROR] Quarkus Extension Dependency Verification Error
[ERROR] Deployment artifact io.quarkus:quarkus-reactive-datasource-deployment::jar:999-SNAPSHOT was found to be missing dependencies on Quarkus extension artifacts marked with '-' below:
[ERROR] + io.quarkus:quarkus-reactive-datasource-deployment::jar:999-SNAPSHOT
[ERROR] +     io.quarkus:quarkus-reactive-datasource::jar:999-SNAPSHOT
[ERROR] +     io.quarkus:quarkus-core-deployment::jar:999-SNAPSHOT
[ERROR] +     io.quarkus:quarkus-arc-deployment::jar:999-SNAPSHOT
[ERROR] +     io.quarkus:quarkus-datasource-deployment::jar:999-SNAPSHOT
[ERROR] -     io.quarkus:quarkus-vertx-core-deployment::jar:999-SNAPSHOT
[ERROR] -         io.quarkus:quarkus-netty-deployment::jar:999-SNAPSHOT
```

**Review by geoand — APPROVED** [review]
Very useful, nice work!

File: extensions/panache/hibernate-orm-rest-data-panache/deployment/pom.xml
Comment (geoand) [inline extensions/panache/hibernate-orm-rest-data-panache/deployment/pom.xml:19]: @gytis can you make sure this is correct?

**Review by stuartwdouglas — APPROVED** [review]

**Comment by loicmathieu** [comment]
@aloubyansky 

> if a runtime artifact depends on another extension's runtime artifact then the deployment artifact depends on the corresponding deployment artifact of the other extension.

Are you sure this is a good thing?

I don't remember precisely but I'm sure I already depends in an extension of the runtime dependency on another one without depending on it's deployment module. The deployment module is resmonsible for triggering build steps for an extension, sometime we don't want this for another extension but still uses some classes from the runtime module (think multiple extensions on the same functionality ...). I agree that in this case a common module is better but I still afraid that this change might impact some existing extension in an un-expected way ...

File: extensions/panache/hibernate-orm-rest-data-panache/deployment/pom.xml
Comment (gytis) [reply extensions/panache/hibernate-orm-rest-data-panache/deployment/pom.xml:19]: These two dependencies just swapped places, so no harm there.

File: extensions/panache/rest-data-panache/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (gytis) [inline extensions/panache/rest-data-panache/runtime/src/main/resources/META-INF/quarkus-extension.yaml:1]: Why do we need this file? This module cannot be used on its own and only contains the common classes that are used by the Hibernate ORM REST Data extension (and the future Mongo alternative).

**Comment by aloubyansky** [comment]
@loicmathieu sorry, can you give me an example, e.g. E1.rt depends on E2.rt, etc.

This is not new! There is no change except the check.

File: extensions/panache/rest-data-panache/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (aloubyansky) [reply extensions/panache/rest-data-panache/runtime/src/main/resources/META-INF/quarkus-extension.yaml:1]: First of all, it's not exposed as an extension  to the users. It is in fact not required in this case. I just added it for consistency since the project is following the extension project structure.

File: extensions/panache/rest-data-panache/runtime/pom.xml
Comment (gytis) [inline extensions/panache/rest-data-panache/runtime/pom.xml:50]: I'm not sure this is needed. This module is not exposed as an extension.

**Comment by loicmathieu** [comment]
I'm on PTO so hard to give an example today.

And there is changes as you add deployment modules to existing extensions right? 

I'll try to find an example tomorrow.

**Comment by aloubyansky** [comment]
@loicmathieu if your runtime artifact depends on another extension's runtime artifact but your deployment artifact does not depend on the corresponding deployment artifact of the other extension, that would be a special/advanced case. Please, do let me know if we have this somewhere, as I am not aware of this. In this case you can simply disable the check in the runtime module by adding `skipExtensionValidation` system property or Maven plugin option.
In the majority of cases, however, this will help to catch misconfigurations early.

BTW, somewhat related, given that you are testing performance differences between versions and we touched upon the resolver a couple of times, this could make a noticeable difference https://github.com/quarkusio/quarkus/pull/10014 Although the resolution part is not included into the time reported by Quarkus during the boot.

**Comment by aloubyansky** [comment]
@gytis I removed the extension metadata from the runtime artifact of the rest-data-panache.
@loicmathieu I'll be waiting for your approval.

File: extensions/security-jpa/deployment/pom.xml
Comment (gsmet) [inline extensions/security-jpa/deployment/pom.xml:85]: Minor but if it's really necessary, we need to move it up with the other non-test dependencies.

**Review by gsmet — APPROVED** [review]
I think this should go in.

@loicmathieu if you really have this case, then that means that we need another common shared module.

**Comment by gsmet** [comment]
I will merge that one. If the check is deemed harmful, we can still relax it later but seems better to have it in CR1 and see if problems occur.

**Comment by loicmathieu** [comment]
@aloubyansky I didn't had a lot of time to give an example, so I'm a little late, sorry for that.

I just found it the hard way : my Elasticsearch PR is failing and will need to be re-think.

Elasticsearch has two clients: the low level and the high level. I re-use in the high level some code from the runtime dependency of the low level but I don't want to include the deployment artifact as I don't want the deployment steps of the low level being run.

 We have a common elasticsearch module but it's shared with hibernate-search so I don't want to put a lot of stuff in it (it mostly contains dependency management, health check and native image substitution).

It worked perfectly but now that I must use the deployment module of each dependent extension, I had to include the deployment module of the low level client inside the high level client and, even if I can make it works, there will be duplicate steps and it will produce some CDI beans that I don't need.

So now, I need to create another common extension (I have no idea how to name it) between my low-level and my hight-level client.

By the way, having only the runtime artifact of a dependency is also a trick to include some functionality of an extension in another without having the dependency listed in the "features" of an application. For example you add the jackson deploymement dependency inside the Funqy extension so now, when a user add the Funqy extension, it will see a "jackson" feature that he didn't add (and didn't use). If he use this in conjonction with resteasy-jsonb he may not understand why both jackson and jsonb are listed ...

**Comment by gsmet** [comment]
@loicmathieu maybe we can talk together about it early next week?

**Comment by loicmathieu** [comment]
@gsmet OK for me, it was just an informal comment as I raise some concern about this PR and Alexey ask for more info but I didn't had time to give them.

**Pull request merged**
